homeassistant/components/pioneer/media_player.py

```diff
@@ -109,24 +115,49 @@ def telnet_request(cls, telnet, command, expected_prefix):
 
     def telnet_command(self, command):
         """Establish a telnet connection and sends command."""
-        try:
+        tries = MAX_TRIES
+        while tries > 0:
+            tries -= 1
             try:
-                telnet = telnetlib.Telnet(self._host, self._port, self._timeout)
-            except (ConnectionRefusedError, OSError):
-                _LOGGER.warning("Pioneer %s refused connection", self._name)
-                return
-            telnet.write(command.encode("ASCII") + b"\r")
-            telnet.read_very_eager()  # skip response
-            telnet.close()
-        except telnetlib.socket.timeout:
-            _LOGGER.debug("Pioneer %s command %s timed out", self._name, command)
+                try:
+                    telnet = telnetlib.Telnet(self._host, self._port, self._timeout)
```

Comment:
The device specific interface code and data parsing needs to be extracted to a standalone library published on PyPI.

https://developers.home-assistant.io/docs/creating_component_code_review#4-communication-with-devicesservices